rskj-core/src/main/java/co/rsk/db/importer/provider/BootstrapFileHandler.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * This file is part of RskJ
+ * Copyright (C) 2019 RSK Labs Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package co.rsk.db.importer.provider;
+
+import co.rsk.db.importer.BootstrapImportException;
+import co.rsk.db.importer.BootstrapURLProvider;
+import co.rsk.db.importer.provider.index.data.BootstrapDataEntry;
+import org.bouncycastle.util.encoders.Hex;
+import org.ethereum.crypto.HashUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.*;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.security.SecureRandom;
+import java.util.*;
+
+public class BootstrapFileHandler {
+
+    private static final Logger logger = LoggerFactory.getLogger(BootstrapFileHandler.class);
+    private static final String BOOTSTRAP_DATA_NAME = "bootstrap-data.zip";
+    private static final String EXTRACTED_PATH = "export";
+
+    private final BootstrapURLProvider bootstrapUrlProvider;
+    private final Unzipper unzipper;
+    private Path tempPath;
+
+    public BootstrapFileHandler(
+            BootstrapURLProvider bootstrapUrlProvider,
+            Unzipper unzipper) {
+        this.bootstrapUrlProvider = bootstrapUrlProvider;
+        this.unzipper = unzipper;
+    }
+
+    // for testing purposes these 3 methods are splitted into 3 public items
+    public void setTempDirectory() {
+        this.tempPath = getTempDirectory();
+    }
+
+    public Path getBootstrapPath() {
+        return tempPath.resolve(EXTRACTED_PATH);
+    }
+
+    public void retrieveAndUnpack(Map<String, BootstrapDataEntry> selectedEntries) {
+        Path zipPath = tempPath.resolve(BOOTSTRAP_DATA_NAME);
+
+        List<String> publicKeys = new ArrayList<>(selectedEntries.keySet());
+        String publicKey = publicKeys.get(new SecureRandom().nextInt(publicKeys.size()));
+        String expectedHash = selectedEntries.get(publicKey).getHash();
+
+        String dbSuffix = selectedEntries.get(publicKey).getDb();
+        URL dbURL = bootstrapUrlProvider.getFullURL(dbSuffix);
+
+        downloadFile(zipPath, dbURL);
```

Comment:
Given that the file is currently being loaded entirely on memory for decoding. Instead of storing the zip file on disk we could have it only on ram.
Or storing it in disk and in future iterations decode it linearly to avoid loading the full file on ram.

Just food for thought.